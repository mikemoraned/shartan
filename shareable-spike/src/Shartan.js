import * as React from "react";
import tartan from 'tartan';
import 'tartan-schema';
import 'tartan-processing';
import 'tartan-render';

const schema = tartan.schema.extended;

function errorHandler(error, data, severity) {
    console.log(error, data, severity);
}

const parse = schema.parse({
    errorHandler: errorHandler,
    transformSyntaxTree: tartan.transform.optimize()
});

const format = schema.format({
    defaultColors: schema.colors,
    includeUnusedColors: false,
    includeDefaultColors: true
});


class Shartan extends React.Component {

    constructor(props) {
      super(props);

      this.state = {
          sett: parse('K18 DO5 LY6 DO5 G9 LN16'),
          dimensions : {
              width: props.width,
              height: props.width
          }
      };
    }

    componentDidMount() {
        this.updateCanvas();
    }

    componentDidUpdate() {
        this.updateCanvas();
    }

    updateCanvas() {
        const { sett } = this.state;

        const render = tartan.render.canvas(sett, {
            defaultColors: schema.colors,
            transformSyntaxTree: tartan.transform.flatten()
        });

        render(this.refs.canvas, { x: 0, y: 0});
    }

    render() {
        return (
            <canvas ref="canvas"
                    width={this.state.dimensions.width}
                    height={this.state.dimensions.height}
            />
        );
    }
}

export default Shartan;
