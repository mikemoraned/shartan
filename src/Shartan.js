import * as React from "react";
import { ColorMap } from "./ColorMap";
import { Sett, Repeat, Reverse } from "./Sett";

class Shartan extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            sett: this.parseSett(props),
            colorMap: new ColorMap(),
            dimensions : {
                width: props.width,
                height: props.width
            }
        };
    }

    parseSett(props) {
        return props.sett == null ? Sett.parse("K4 R24 K24 Y4") : Sett.parse(props.sett);
    }

    componentDidMount() {
        this.updateCanvas();
    }

    componentWillReceiveProps(nextProps) {
        this.setState({
            sett: this.parseSett(nextProps)
        });
    }

    componentDidUpdate() {
        this.updateCanvas();
    }

    updateCanvas() {
        const { width, height } = this.state.dimensions;
        const { sett } = this.state;

        const ctx = this.refs.canvas.getContext('2d');
        ctx.clearRect(0, 0, width, height);

        this.drawWarp(ctx, sett.pivoted([Repeat, Reverse, Repeat, Reverse]));
        this.drawWeft(ctx, sett.pivoted([Reverse, Repeat, Reverse, Repeat]));
    }

    drawWarp(context, threadCounts) {
        let lastEndX = 0;
        threadCounts.mapPercentages((percentage, threadCount) => {
            const width = this.state.dimensions.width * (percentage / 100);
            const height = this.state.dimensions.height;

            context.fillStyle = this.state.colorMap.colorFor(threadCount.colorName);
            context.fillRect(lastEndX, 0, width, height);

            lastEndX = lastEndX + width;
        })
    }

    drawWeft(context, threadCounts) {
        const prevGlobalAlpha = context.globalAlpha;
        context.globalAlpha = 0.5;
        let lastEndY = 0;
        threadCounts.mapPercentages((percentage, threadCount) => {
            const width = this.state.dimensions.width;
            const height = this.state.dimensions.height * (percentage / 100);

            context.fillStyle = this.state.colorMap.colorFor(threadCount.colorName);
            context.fillRect(0, lastEndY, width, height);

            lastEndY = lastEndY + height;
        });
        context.globalAlpha = prevGlobalAlpha;
    }

    render() {
        return (
            <canvas ref="canvas"
                    width={this.state.dimensions.width}
                    height={this.state.dimensions.height}/>
        );
    }
}

export default Shartan;