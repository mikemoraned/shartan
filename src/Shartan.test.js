import Shartan from "./Shartan";
import React from 'react';
import renderer from 'react-test-renderer';

test('renders default', () => {
    const tree = renderer.create(
        <Shartan width={400}/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
