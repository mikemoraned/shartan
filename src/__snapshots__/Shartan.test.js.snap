// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default 1`] = `
<div
  style={
    Object {
      "height": "400px",
      "position": "absolute",
      "top": "0px",
      "width": "400px",
    }
  }
>
  <div
    style={
      Object {
        "height": "400px",
        "position": "absolute",
        "top": "0px",
        "width": "400px",
      }
    }
  >
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "float": "left",
          "height": "100%",
          "width": "10.714285714285714%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "10.714285714285714%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "10.714285714285714%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "float": "left",
          "height": "100%",
          "width": "10.714285714285714%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "float": "left",
          "height": "100%",
          "width": "10.714285714285714%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "10.714285714285714%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "10.714285714285714%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "float": "left",
          "height": "100%",
          "width": "10.714285714285714%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
  </div>
  <div
    style={
      Object {
        "height": "400px",
        "position": "absolute",
        "top": "0px",
        "width": "400px",
      }
    }
  >
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "10.714285714285714%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "height": "10.714285714285714%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "height": "10.714285714285714%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "10.714285714285714%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "10.714285714285714%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "height": "10.714285714285714%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "height": "10.714285714285714%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "10.714285714285714%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
  </div>
</div>
`;

exports[`renders when sett specified 1`] = `
<div
  style={
    Object {
      "height": "400px",
      "position": "absolute",
      "top": "0px",
      "width": "400px",
    }
  }
>
  <div
    style={
      Object {
        "height": "400px",
        "position": "absolute",
        "top": "0px",
        "width": "400px",
      }
    }
  >
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "float": "left",
          "height": "100%",
          "width": "10.714285714285714%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "4.017857142857143%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "float": "left",
          "height": "100%",
          "width": "6.696428571428571%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "float": "left",
          "height": "100%",
          "width": "6.696428571428571%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "4.017857142857143%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "float": "left",
          "height": "100%",
          "width": "10.714285714285714%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "float": "left",
          "height": "100%",
          "width": "10.714285714285714%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "4.017857142857143%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "float": "left",
          "height": "100%",
          "width": "6.696428571428571%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "float": "left",
          "height": "100%",
          "width": "6.696428571428571%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "4.017857142857143%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "float": "left",
          "height": "100%",
          "width": "10.714285714285714%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "float": "left",
          "height": "100%",
          "width": "1.7857142857142858%",
        }
      }
    >
       
    </div>
  </div>
  <div
    style={
      Object {
        "height": "400px",
        "position": "absolute",
        "top": "0px",
        "width": "400px",
      }
    }
  >
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "height": "6.696428571428571%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "4.017857142857143%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "height": "10.714285714285714%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "height": "10.714285714285714%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "4.017857142857143%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "height": "6.696428571428571%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "height": "6.696428571428571%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "4.017857142857143%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "height": "10.714285714285714%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "height": "10.714285714285714%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "4.017857142857143%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FF0000",
          "height": "6.696428571428571%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#FFFF00",
          "height": "1.7857142857142858%",
          "opacity": "0.5",
          "width": "100%",
        }
      }
    >
       
    </div>
  </div>
</div>
`;
